applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp

```diff
@@ -0,0 +1,559 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/IGAStructuralMechanicsApplication/license.txt
+//
+//  Main authors:    Tobias Teschemacher
+//
+
+
+// System includes
+#include "includes/define.h"
+#include "utilities/math_utils.h"
+
+// External includes
+
+// Project includes
+#include "custom_elements/base_discrete_element.h"
+
+#include "iga_structural_mechanics_application.h"
+#include "iga_structural_mechanics_application_variables.h"
+
+#include "geometries/geometry.h"
+
+namespace Kratos
+{
+    void BaseDiscreteElement::Initialize()
+    {
+        KRATOS_TRY
+
+        //Constitutive Law initialisation
+        if (mConstitutiveLawVector.size() != 1)
+            mConstitutiveLawVector.resize(1);
+
+        InitializeMaterial();
+
+        KRATOS_CATCH("")
+    }
+
+    /***********************************************************************************/
+    /***********************************************************************************/
+    void BaseDiscreteElement::InitializeMaterial()
+    {
+        KRATOS_TRY
+
+            if (GetProperties()[CONSTITUTIVE_LAW] != nullptr) {
+                mConstitutiveLawVector[0] = GetProperties()[CONSTITUTIVE_LAW]->Clone();
+                mConstitutiveLawVector[0]->InitializeMaterial(GetProperties(),
+                    GetGeometry(),
+                    GetValue(SHAPE_FUNCTION_VALUES)
+                );
+            }
+            else
+                KRATOS_ERROR << "A constitutive law needs to be specified for the element with ID " << this->Id() << std::endl;
+
+        KRATOS_CATCH("");
+    }
+
+    /***********************************************************************************/
+    /***********************************************************************************/
+    void BaseDiscreteElement::ResetConstitutiveLaw()
+    {
+        KRATOS_TRY
+
+        if (GetProperties()[CONSTITUTIVE_LAW] != nullptr) {
+                mConstitutiveLawVector[0]->ResetMaterial(GetProperties(),
+                    GetGeometry(),
+                    GetValue(SHAPE_FUNCTION_VALUES)
+                );
+        }
+
+        KRATOS_CATCH("")
+    }
+
+    /***********************************************************************************/
+    /***********************************************************************************/
+    void BaseDiscreteElement::EquationIdVector(
+        EquationIdVectorType& rResult,
+        ProcessInfo& rCurrentProcessInfo
+    )
+    {
+        KRATOS_TRY;
+
+        const unsigned int number_of_control_points = GetGeometry().size();
```

Comment:
I would avoid the usage of unsigned datatypes since they have no benefits and are highly error prone

Personally I would also drop the `const` if it is clear for the compiler and the user

Reply:
for me it is often not clear whether sth is const or not
I usually but it where possible, since it can prevent errors